File: Sources/Sentry/SentrySdkPackage.m
#import "SentrySdkPackage.h"
#import "SentryMeta.h"
#import <Foundation/Foundation.h>

typedef NS_ENUM(NSUInteger, SentryPackageManagerOption) {
    SentrySwiftPackageManager,
    SentryCocoaPods,
    SentryCarthage,
    SentryPackageManagerUnkown
};

/**
 * This is required to identify the package manager used when installing sentry.
 */
#if SWIFT_PACKAGE
static SentryPackageManagerOption SENTRY_PACKAGE_INFO = SentrySwiftPackageManager;
#elif COCOAPODS
static SentryPackageManagerOption SENTRY_PACKAGE_INFO = SentryCocoaPods;
#elif CARTHAGE_YES
// CARTHAGE is a xcodebuild build setting with value `YES`, we need to convert it into a compiler
// definition to be able to use it.
static SentryPackageManagerOption SENTRY_PACKAGE_INFO = SentryCarthage;
#else
static SentryPackageManagerOption SENTRY_PACKAGE_INFO = SentryPackageManagerUnkown;
#endif

NS_ASSUME_NONNULL_BEGIN

@interface SentrySdkPackage ()

@property (nonatomic) SentryPackageManagerOption packageManager;

@end

@implementation SentrySdkPackage

- (instancetype)initWithName:(NSString *)name andVersion:(NSString *)version
{
    if (self = [super init]) {
        _name = name ?: @"";
        _version = version ?: @"";
        _packageManager = SENTRY_PACKAGE_INFO;
    }

    return self;
}

- (nullable instancetype)initWithDict:(NSDictionary *)dict
{
    NSString *name = @"";
    NSString *version = @"";

    if (![dict[@"name"] isKindOfClass:[NSString class]]) {
        return nil;
    } else {
        name = dict[@"name"];
    }

    if (![dict[@"version"] isKindOfClass:[NSString class]]) {
        return nil;
    } else {
        version = dict[@"version"];
    }

    return [self initWithName:name andVersion:version];
}

- (NSDictionary<NSString *, id> *)serialize
{
    return @{
        @"name" : self.name,
        @"version" : self.version,
    };
}

+ (nullable NSString *)getSentrySDKPackageName
{
    switch (SENTRY_PACKAGE_INFO) {
    case SentrySwiftPackageManager:
        return @"spm:getsentry/%@";
    case SentryCocoaPods:
        return @"cocoapods:getsentry/%@";
    case SentryCarthage:
        return @"carthage:getsentry/%@";
    default:
        return nil;
    }
}

+ (nullable SentrySdkPackage *)getSentrySDKPackage
{

    if (SENTRY_PACKAGE_INFO == SentryPackageManagerUnkown) {
        return nil;
    }

    NSString *name = [SentrySdkPackage getSentrySDKPackageName];
    if (nil == name) {
        return nil;
    }

    return [[SentrySdkPackage alloc] initWithName:name andVersion:SentryMeta.versionString];
}

@end

NS_ASSUME_NONNULL_END

File: Sources/Swift/Helper/SentrySdkPackage.swift
@_implementationOnly import _SentryPrivate
import Foundation

/// This is required to identify the package manager used when installing sentry.
private enum SentryPackageManagerOption: UInt {
    case swiftPackageManager = 0
    case cocoaPods = 1
    case unknown = 2
}

#if SWIFT_PACKAGE
private var SENTRY_PACKAGE_INFO: SentryPackageManagerOption = .swiftPackageManager
#elseif COCOAPODS
private var SENTRY_PACKAGE_INFO: SentryPackageManagerOption = .cocoaPods
#else
private var SENTRY_PACKAGE_INFO: SentryPackageManagerOption = .unknown
#endif

@objc
@_spi(Private) public final class SentrySdkPackage: NSObject {

    private static func getSentrySDKPackageName(_ packageManager: SentryPackageManagerOption) -> String? {
        switch packageManager {
        case .swiftPackageManager:
            return "spm:getsentry/\(SentryMeta.sdkName)"
        case .cocoaPods:
            return "cocoapods:getsentry/\(SentryMeta.sdkName)"
        case .unknown:
            // We don't know if the user installed Sentry with Xcode, manually or Carthage using the prebuild xcframework
            return nil
        }
    }

    private static func getSentrySDKPackage(_ packageManager: SentryPackageManagerOption) -> [String: String]? {
        if packageManager == .unknown {
            return nil
        }

        guard let name = getSentrySDKPackageName(packageManager) else {
            return nil
         }

        return ["name": name, "version": SentryMeta.versionString]
    }

    @objc
    public static func global() -> [String: String]? {
        return getSentrySDKPackage(SENTRY_PACKAGE_INFO)
    }

    #if SENTRY_TEST || SENTRY_TEST_CI
    @objc
    public static func setPackageManager(_ manager: UInt) {
        SENTRY_PACKAGE_INFO = SentryPackageManagerOption(rawValue: manager) ?? .unknown
    }

    @objc
    public static func resetPackageManager() {
        SENTRY_PACKAGE_INFO = .unknown
    }
    #endif
}
